source/Modules/Controller/connection.js

```diff
@@ -0,0 +1,203 @@
+import qs from 'query-string';
+import extension from 'extensionizer';
+import ERRORS from '@background/errors';
+import PlugController from '@psychedelic/plug-controller';
+import { validatePrincipalId } from '@shared/utils/ids';
+import { CONNECTION_STATUS } from '@shared/constants/connectionStatus';
+import {
+  getApps,
+  setApps,
+  getApp,
+  removeApp
+} from '../storageManager';
+import SIZES from '../../Pages/Notification/components/Transfer/constants';
+
+export class ConnectionModule {
+  constructor(backgroundController, secureController, keyring) {
+    this.keyring = keyring;
+    this.secureController = secureController;
+    this.backgroundController = backgroundController;
+  }
+
+  // Utils
+  #getHandlerObjects() {
+    return [this.#isConnected(), this.#disconnect(), this.#requestConnect()];
+  }
+
+  #secureWrapper({ args, handlerObject }) {
+    return this.secureController(
+      args[0].callback,
+      async () => {
+        handlerObject.handler(...args);
+      }
+    );
+  }
+
+  async #fetchCanistersInfo(whitelist) {
+    if (whitelist && whitelist.length > 0) {
+      const canistersInfo = await Promise.all(
+        whitelist.map(async (id) => {
+          let canisterInfo = { id };
+
+          try {
+            const fetchedCanisterInfo = await PlugController.getCanisterInfo(id);
+            canisterInfo = { id, ...fetchedCanisterInfo };
+          } catch (error) {
+            /* eslint-disable-next-line */
+            console.error(error);
+          }
+
+          return canisterInfo;
+        }),
+      );
+
+      const sortedCanistersInfo = canistersInfo.sort((a, b) => {
+        if (a.name && !b.name) return -1;
+        return 1;
+      });
+
+      return sortedCanistersInfo;
+    }
+
+    return [];
+  }
+
+  // Handlers
+  #isConnected() {
+    return {
+      methodName: 'isConnected',
+      handler: async (opts, url) => {
+        const { callback } = opts;
+
+        getApp(this.keyring.currentWalletId.toString(), url, (app) => {
+          if (app) {
+            callback(null, app.status === CONNECTION_STATUS.accepted);
+            return;
+          }
+
+          callback(null, false);
+        });
+      },
+    };
+  };
+
+  #disconnect() {
+    return {
+      methodName: 'disconnect',
+      handler: async (opts, url) => {
+        removeApp(this.keyring.currentWalletId.toString(), url, (removed) => {
```

Comment:
I'd add optional chaining everywhere to prevent weird crashes on the keyring

Reply:
Sounds like a good measure, I'll make sure I add optional chaining on the whole module